### DISCUSSION TITLE
dble中执行语句慢，怎么排查

### DISCUSSION THREAD

**Discussion by PanternBao** [actiontech/dble#2958] [Q&A] [answered]
一段很简单的SQL，但dble查询和单库查询执行时间对比，相差n秒，dble执行查询慢在哪里？怎么排查？

**Comment by Dinosauria** [comment, marked as answer]
排查方向：
1. 使用trace https://actiontech.github.io/dble-docs-cn/2.Function/2.21_query_trace.html
2. 开启dble的慢日志 https://actiontech.github.io/dble-docs-cn/2.Function/2.20_slow_query_log.html （@@slow_query.time=200 慢查询统计阈值，单位是毫秒）
 
找出具体哪个环节耗时较多

**Comment by ylinzhu** [comment]
## 前置知识
需要简单了解下dble一条sql的内部流程
![image](https://user-images.githubusercontent.com/46126022/143521066-dfe01cc9-c5d8-4d4e-875d-6bbc03ce414e.png)
需要对中间执行时间过长的状态进行分析
那么通常情况主要有什么状态会导致dble中执行语句慢呢
* dble从mysql接收数据慢
* dble在处理复杂查询中进行合并等操作
* 其他sql阻塞了这条sql

慢日志中可以体现上述步骤的各个阶段耗时。

各个阶段的耗时说明：
1. T2-T3的耗时表示 dble 和 mysql节点的最快交互时长。同2
2. T3-T5的耗时表示 dble 和 mysql节点的最慢交互时长。如果这个部分耗时和实际sql 在 mysql 中执行时长相差比较大，这里有可能是网络原因，可能是sql在mysql执行耗时长（比如，有其他sessioin开启事务未提交，而这条语句正好需要等待他的提交，或者是mysql中数据多等情况）
3. T5-T6的耗时表示 dble中处理中间结果集的时间。

举例
存在类似sql
`table_a a left join table_b b on a.col_1 =b .col_1 left join table_c c on a.col_2 =c.col_2 where a.col =xxx`
最终结果虽然可能只有几百条数据，但是在dble中处理的结果可能会有很多，可以通过执行计划直观的看到真实的下发语句
该sql的查询计划类似
a表带条件下发，结果集大约数百
b表大表，全数据拉取
c表小表，全数据拉取
使用某些工具时，当你下发一条sql，它可能会先下发其他sql到dble，然后在下发你执行的sql，所以如果存在之前语句执行慢也会出现看起来sql执行慢的情况



文档中有使用dble的慢日志说明 [如何开启慢日志](https://actiontech.github.io/dble-docs-cn/2.Function/2.20_slow_query_log.html )
